File: src/main/java/com/shojabon/man10fishing/dataClass/FishFactor.java
package com.shojabon.man10fishing.dataClass;

import com.shojabon.man10fishing.annotations.FishFactorDefinition;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

public abstract class FishFactor {

    Fish fish;

    public FishFactor(Fish fish){
        this.fish = fish;
    }

    public boolean fishEnabled(Fish fish, Player fisher, FishingRod rod){
        return true;
    }

    public float rarityMultiplier(Fish fish, float currentMultiplier, Player fisher, FishingRod rod){
        return 1f;
    }

    public FishFactorDefinition getDefinition(){
        if(!this.getClass().isAnnotationPresent(FishFactorDefinition.class)) return null;
        return this.getClass().getAnnotation(FishFactorDefinition.class);
    }

    public void onFish(Fish fish, Player fisher, FishingRod rod){}



}

File: src/main/java/com/shojabon/man10fishing/dataClass/FishFactor.kt
package com.shojabon.man10fishing.dataClass

import org.bukkit.entity.Player
import com.shojabon.man10fishing.annotations.FishFactorDefinition
import org.bukkit.Location

abstract class FishFactor(var fish: Fish) {
    open fun fishEnabled(fish: Fish, fisher: Player, rod: FishingRod, hookLocation:Location): Boolean {
        return true
    }

    open fun rarityMultiplier(fish: Fish, currentMultiplier: Float, fisher: Player, rod: FishingRod): Float {
        return currentMultiplier
    }

    open fun onFish(fish: Fish, parameter: FishParameter, fisher: Player, rod: FishingRod) {
    }

    val definition: FishFactorDefinition?
        get() = if (!this.javaClass.isAnnotationPresent(FishFactorDefinition::class.java)) null else this.javaClass.getAnnotation(FishFactorDefinition::class.java)

}